plugins/amdgpu/tests/test_topology_remap.c

```diff
@@ -0,0 +1,985 @@
+ *
+ * Test 7: Checkpoint node has 8 GPUs (4 in XGMI hive, 4 without XGMI bridge)
+ *      Restore node has 8 GPUs (2 XGMI hives)
+ *      Restore should succeed - user application will not take advantage of 2 XGMI bridge
+ *      EXPECT: SUCCESS
+ *
+ * Test 8: Checkpoint node has 5 GPUs (3 GPU's with P2P-PCIe links)
+ *     Restore node has 5 GPUs (4 GPU's with P2P-PCIe links)
+ *      EXPECT: SUCCESS
+ *
+ **************************************************************************************************/
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <errno.h>
+#include "common/list.h"
+
+#include "amdgpu_plugin_topology.h"
+
+
+#define pr_err(...) fprintf(stdout, "ERR:" __VA_ARGS__)
+#define pr_info(...) fprintf(stdout, "INFO:" __VA_ARGS__)
+#define pr_debug(...) fprintf(stdout, "DBG:" __VA_ARGS__)
+
+int verify_maps(const struct device_maps *maps, uint32_t num_cpus, uint32_t num_gpus)
+{
+        struct id_map *map;
+
+        list_for_each_entry(map, &maps->cpu_maps, listm) {
+                if (num_cpus-- == 0) {
+                        pr_err("Results had more mappings than number of CPUs\n");
+                        return -EINVAL;
+                }
+        }
+        if (num_cpus > 0) {
+                pr_err("Results did not map all CPUs\n");
+                return -EINVAL;
+        }
+
+        list_for_each_entry(map, &maps->gpu_maps, listm) {
+                if (num_gpus-- == 0) {
+                        pr_err("Results had more mappings than number of GPUs\n");
+                        return -EINVAL;
+                }
+        }
+        if (num_gpus > 0) {
+                pr_err("Results did not map all GPUs\n");
+                return -EINVAL;
+        }
+	return 0;
+}
+
+int test_0(void)
+{
+        int ret = 0;
+        struct device_maps maps;
+
+        struct tp_node *node_gpus[8];
+        struct tp_node *node_cpus[2];
+
+        struct tp_system tp_src = { 0 };
+        struct tp_system tp_dest = { 0 };
+
+        /* Fill src struct */
+        topology_init(&tp_src);
+        tp_src.parsed = true;
+
+        node_cpus[0] = sys_add_node(&tp_src, 0, 0);
+        node_cpus[0]->cpu_cores_count = 1;
+        node_cpus[1] = sys_add_node(&tp_src, 1, 0);
+        node_cpus[1]->cpu_cores_count = 1;
+
+        for (int i = 0; i < 4; i++) {
+                node_gpus[i] = sys_add_node(&tp_src, i + 2, 0xA000 + i);
+                node_gpus[i]->device_id = 0xD000;
+                node_add_iolink(node_gpus[i], TOPO_IOLINK_TYPE_PCIE, i & 1);
+                node_add_iolink(node_cpus[i & 1], TOPO_IOLINK_TYPE_PCIE, i + 2);
+        }
+        for (int i = 0; i < 4; i++) {
+                for (int j = 1; j < 4; j++)
+                        node_add_iolink(node_gpus[i], TOPO_IOLINK_TYPE_XGMI, ((i + j) % 4) + 2);
+                for (int j = 4; j < 8; j++)
+                        node_add_iolink(node_gpus[i], TOPO_IOLINK_TYPE_PCIE, ((i + j) % 4) + 6);
```

Comment:
>This adds links to nodes that haven't been created yet. Does that work?
Yes, Node is created 8 lines above: 
Line 139: node_gpus[i] = sys_add_node(&tp_src, i + 2, 0xA000 + i);

>Also, the %4 + 6 is confusing here. I think this would be clearer like this:
ACK

Reply:
> Yes, Node is created 8 lines above:

No. You're creating links to node 6-9 (the second hive) here. Those are created below in the "for (int i = 4; i < 8; i++)" loop.

Reply:
>>Yes, Node is created 8 lines above:
>No. You're creating links to node 6-9 (the second hive) here. Those are created below in the "for (int i = 4; i < 8; i++)" loop.
That is ok. We can create links to node that do not exist yet because the links are validated later during topology_determine_iolinks(). It is only the source node that has to exist at the time the link is created.